---
title: "Characterizing TCGA EMT-high tumours"
author: "Zhiyuan Hu"
date: "28/11/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = F)

suppressPackageStartupMessages(library(edgeR))
suppressPackageStartupMessages(library(limma))

suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(Biobase))
                               
eset <- readRDS("../../scFT-paper_rds/20191009_esetTCGA_afterDeconvolution.rds")
```


## Markers of EMT-high

```{r DE-analysis-EMT-lowVShigh}
eset$emt_strata <- NA
eset$emt_strata[eset$emt_new >= quantile(eset$emt_new, 2/3)] <- "High"
eset$emt_strata[eset$emt_new <= quantile(eset$emt_new, 1/3)] <- "Low"

eset$emt_strata <- factor(eset$emt_strata, levels = c("Low","High"))


dge <- DGEList(counts=eset@assayData$exprs[,eset$emt_strata %in% c("Low","High")])
A <- rowSums(dge$counts) # filter genes
dge <- dge[A > 10,, keep.lib.sizes=FALSE]
dge <- calcNormFactors(dge)


group <- eset$emt_strata[ eset$emt_strata %in% c("Low","High")]
design <- model.matrix(~0+group) # model by EMT groups

# limma
v <- voom(dge, design, plot = F)
fit <- lmFit(v, design)

cont.matrix <- makeContrasts(groupHigh-groupLow,
                             levels = design)
fit <- contrasts.fit(fit, cont.matrix)
fit <- eBayes(fit)

marker_EMThigh <- topTable(fit,number = Inf, p.value = 0.05, lfc = 1, coef = 1)
marker_EMThigh$gene <- rownames(marker_EMThigh)
marker_EMThigh <- marker_EMThigh[order(marker_EMThigh$logFC, decreasing = T),]

# write.csv(marker_EMThigh[marker_EMThigh$logFC >= 1,],"results/TableS7_markers_EMThigh20190215.csv")
# write.csv(marker_EMThigh,"../other_plots/markers_EMThigh20191008.csv")
```

```{r print-head-DElist}
knitr::kable(head(marker_EMThigh))
```


```{r vlnplot_EMT_TFs}
df_plot <- data.frame(exprs = c(eset@assayData$exprs["TWIST1",],eset@assayData$exprs["TWIST2",],eset@assayData$exprs["SNAI2",]), 
                      group = rep(eset$emt_strata,3), 
                      gene = rep(c("TWIST1","TWIST2","SNAI2"), each = ncol(eset)))
df_plot <- na.omit(df_plot)
df_plot$exprs <- log1p(df_plot$exprs)
df_plot$group <- paste("EMT-", df_plot$group, sep = "")


ggplot(df_plot, aes(x = group, y = exprs)) + 
  geom_violin() + theme_pubr() + 
  geom_boxplot(width  = 0.1) + 
  geom_jitter(shape=16, alpha = 0.3, position=position_jitter(0.2)) +
  xlab("") + ylab("Expression level") +
  facet_grid(~gene,margins = "am", scales = "free") +
  theme(strip.text.x = element_text(size=12, angle=0, face = "italic"),
        strip.background = element_rect(colour="black", fill="white"))

# ggsave("plots/Fig4_TWIST_SNAI2_violin20190215.pdf", width = 6, height = 3)
```


## miRNA

#### DE analysis of miRNA

```{r DE-analysis-miRNA}
miRNA <- read.delim("../../scFT-paper_rds/TCGA/Xena/miRNA/gdac.broadinstitute.org_OV.Merge_mirnaseq__illuminahiseq_mirnaseq__bcgsc_ca__Level_3__miR_gene_expression__data.Level_3.2016012800.0.0/OV.mirnaseq__illuminahiseq_mirnaseq__bcgsc_ca__Level_3__miR_gene_expression__data.data.txt", sep = "\t",row.names = 1,as.is = T)
# rownames(miRNA) <- miRNA[,1]
miRNA <- miRNA[,miRNA[1,] == "read_count"]
colnames(miRNA) <- gsub(pattern = "[.]", replacement = "-", x = colnames(miRNA))
colnames(miRNA) <- substr(colnames(miRNA), start = 1, stop = 15)

miRNA <- miRNA[,na.omit(match(colnames(eset), colnames(miRNA)))]
miRNA <- miRNA[-1,]

for(itor in 1:ncol(miRNA)){ ## transfer it to numeric
  miRNA[,itor] <- as.numeric(miRNA[,itor])
}

group <- eset$emt_strata[match(colnames(miRNA), colnames(eset))]

dge <- DGEList(counts=miRNA[,!is.na(group)])
A <- rowSums(dge$counts) # filter genes
dge <- dge[A > 10,, keep.lib.sizes=FALSE]
dge <- calcNormFactors(dge)

group <- na.omit(group)
design <- model.matrix(~0+group) # model by EMT groups

# limma
v <- voom(dge, design, plot = F)
fit <- lmFit(v, design)

cont.matrix <- makeContrasts(groupHigh-groupLow, levels = design)
fit <- contrasts.fit(fit, cont.matrix)
fit <- eBayes(fit)

de_genes <- topTable(fit, number = Inf, p.value = 0.05)
de_genes$gene <- rownames(de_genes)
de_genes <- de_genes[order(de_genes$logFC, decreasing = T),]
# write.csv(de_genes, file = "results/miRNA_TCGA/20181211miRNA_DE_EMTHighvsLow.csv")

knitr::kable(head(de_genes))
```

##### Volcano plot

```{r volcano=plot}
## http://bioinformatics.knowledgeblog.org/2011/06/21/volcano-plots-of-microarray-data/

tt <- topTable(fit, number = Inf)
tt$names <- gsub(rownames(tt), pattern = "hsa-mir", replacement = "miR")

##Highlight genes that have an absolute fold change > 2 and a p-value < Bonferroni cut-off
tt$threshold <- as.factor(abs(tt$logFC) > 0.5 & tt$adj.P.Val < 0.05)
tt$threshold2 <- as.factor(abs(tt$logFC) > 1.5 & tt$adj.P.Val < 0.001)
tt$threshold2[tt$names %in% c("miR-200a","miR-200b","miR-200c","miR-141","miR-429")] <- T
tt$threshold2[tt$names %in% c("miR-514-2","miR-514-3","miR-509-2","miR-509-3")] <- F

##Construct the plot object
ggplot(data = tt, aes(x = logFC, y = -log10(P.Value), colour = paste(threshold, threshold2))) +
  geom_point(alpha = 0.4, size = 1.75) +
  xlim(c(-3, 3)) + ylim(c(0, 13)) +
  xlab("log2 fold change") + ylab("-log10 p-value")+ 
  geom_text_repel(data = tt[tt$threshold2 == TRUE,],
                  aes(x = logFC, y = -log10(P.Value), label = names), colour = "black", alpha = 0.75) + 
  theme_pubr() + theme(legend.position = "none")

# ggsave("plots/Fig_miRNA_volcano_plot20190215.png", width = 5, height = 4)
```


## Technical

```{r sessionInfo}
sessionInfo()
```