---
title: "Deconvolution of bulk expression data"
author: "Zhiyuan Hu"
date: "2019/11/25"
output: 
    html_document:
        toc: true
        number_sections: true
        theme: united
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, fig.width = 6, fig.height = 4)

# devtools::install_github('shenorrlab/bseq-sc')
library(bseqsc)

library(MASS)
library(xbioc)
library(e1071) #svm
library(preprocessCore)

library(survminer)
library(survival)
library(powerSurvEpi)
library(survminer)

library(Biobase)
library(SingleCellExperiment)
library(scater)

library(ggbeeswarm)
library(ggrepel)
library(viridis)

library(edgeR)
library(limma)
# library(copynumber)

library(ggplot2)
library(RColorBrewer)
```

# Summary

# Single cell data

```{r read-single-cell-data, include = F}
source("../../scFT-paper_other/modified_functions20190204.R")

# Single cell data - Fresh cells
secretory <- readRDS("../../scFT-paper_rds/20190120Fresh_secretory_9clusters_clincluster.rds")
sceset <- readRDS("../../scFT-paper_rds/20180917Sceset_12clusters.rds")
markers2 <- read.csv("../../scFT-paper_rds/20190120Clincluster_fresh_secretory_9clusters_markers.csv", row.names = 1, as.is = T)
markers_sceset <- read.csv("../../scFT-paper_rds/20180917sceset_markers_12cluster-Zhiyuan’s iMac.csv", row.names = 1, as.is = T)
fresh <- sceset[,sceset$source == "Fresh"]
# rm(secretory)
rm(sceset)
# save.image("../../scFT-paper_rds/20191125background_deconvolution_scRNAseqData.rdata", compress = T)
```

```{r read-single-cell-data-image}
load("../../scFT-paper_rds/20191125background_deconvolution_scRNAseqData.rdata")
```


# Select Markers

We selected the representative marker genes to denote the transcriptomic signatures of the five FTE cell subtypes.

```{r select-markers}
marker.C3 <- markers2[markers2$cluster == "C3" & markers2$logFC > 1.79, ]
marker.C4 <- markers2[markers2$cluster == "C4" & markers2$logFC > 1.5 & markers2$ratio2 < 0.3,]
marker.EMT <- markers2[markers2$cluster == "C8" & markers2$logFC > 2 & markers2$ratio2 < 0.2,]
marker.C10 <- markers2[markers2$cluster == "C10" & markers2$logFC > 2.5 & markers2$ratio2 < 0.2,]
marker.ciliated <- markers_sceset[markers_sceset$cluster == 9 & markers_sceset$logFC > 6 & markers_sceset$ratio2 < 0.1,]

markers <- list(C3 = marker.C3$gene,
                C4 = marker.C4$gene,
                EMT = marker.EMT$gene,
                C10 = marker.C10$gene,
                ciliated = marker.ciliated$gene)
```


## Construct a scRNA-seq dataset for fresh FTE cells

```{r perpare-fresheset}
fresheset <- ExpressionSet(assayData = as.matrix(counts(fresh)),
                           phenoData = new("AnnotatedDataFrame",
                                           data=as.data.frame(fresh@colData)))
fresheset$cell.type <- NA
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C3"]] <- "C3" # differentiated
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C4"]] <- "C4" # KRT17 cluster
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C8"]] <- "EMT"
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C10"]] <- "C10" # cell cycle clusters
fresheset$cell.type[fresheset$clincluster == 9] <- "ciliated"

fresheset <- fresheset[,!is.na(fresheset$cell.type)]
table(fresheset$cell.type)
```

```{r prepare-fresh-SingleCellExperiment_object}
fresh <- fresh[,match(colnames(fresheset), colnames(fresh))]
fresh$cell.type <- fresheset$cell.type
```


# Data

The eset datasets are preprocessed by '20181028Prepare_datasets.rmd'

The signature matrix is actually the mean expression within one cluster after normalisation (non-log-space).

We read the TCGA, AOCS and OXO-PCR preprocessed data from a saved RData file.


```{r read-tcga-data, eval=F}
## TCGA data
eset <- readRDS("../../man_analysis3_20190724/R_deconvolution/rds/20181029TCGA_eset.rds")
exprs_tcga <- readRDS("../../man_analysis3_20190724/R_deconvolution/rds/20181029TCGA_exprs_nonlog.rds")

tcga <- SingleCellExperiment(assays = list(counts = eset@assayData$exprs),
                             colData = as.data.frame(pData(eset)))
logcounts(tcga) <- log2(calculateCPM(tcga) + 1)

eset$stage_dichotomized <- "early"
eset$stage_dichotomized[eset$clinical_stage%in% c("Stage IIIA","Stage IIIB","Stage IIIC","Stage IV")] <- "late"

eset$residual_disease1 <- NA
eset$residual_disease1[eset$tumor_residual_disease %in% c(">20 mm", "11-20 mm") ] <- "residual"
eset$residual_disease1[eset$tumor_residual_disease %in% c("No Macroscopic disease","1-10 mm") ] <- "no_residual"

## AOCS data
toteset <- readRDS("../../man_analysis3_20190724/R_deconvolution/rds/20181029_tothill_eset.rds")
exprs_tot <- readRDS("../../man_analysis3_20190724/R_deconvolution/rds/20181029_tothill_exprs_nonlog.rds")

pData(toteset)$OS <- FALSE
pData(toteset)$OS[pData(toteset)$patient.status == "D"] <- TRUE

toteset$histological.subtype2 <- "ser"
toteset$histological.subtype2[toteset$histological.subtype != "Ser"] <- "non-ser"
toteset$histological.subtype2 <- as.factor(toteset$histological.subtype2)

## OXO-PCR data
oxopcr_pre <- readRDS("../../../OXO_PCR/R/rds/OXO_PCR_Allsamples_2018Oct_PreChemo.rds")
oxopcr_info_pre <- readRDS("../../../OXO_PCR/R/rds/OXO_PCR_Allsamples_2018Oct_PreChemo_sampleInfo.rds")

# save.image("../../scFT-paper_rds/bulkExprData_forDeconvolution20191125.rdata", compress = T)
```

```{r load-bulk-RNAseq-data}
load("../../scFT-paper_rds/bulkExprData_forDeconvolution20191125.rdata")

toteset$summarystage <- NA
toteset$summarystage[toteset$Stage %in% c("I","IA",  "IB" , "IC",  "IIA", "IIB" ,"IIC")] <- "early"
toteset$summarystage[toteset$Stage %in% c("III",  "IIIA", "IIIB", "IIIC", "IV" )] <- "late"

toteset$summarygrade <- NA
toteset$summarygrade[toteset$Grade %in% c(1)] <- "low"
toteset$summarygrade[toteset$Grade %in% c(2,3)] <- "high"

```


# Select genes by PCA in TCGA data

```{r marker_pca_function}
# we removed the marker genes with little dispersion in the bulk RNAseq data
pca_emt <- function(markers){
  markers_in <- intersect(markers, rownames(tcga))
  exprs_mat <- logcounts(tcga)[match(markers_in, rownames(tcga)), ,drop = F]

  ## scale
  exprs_mat <- t(exprs_mat)
  vars <- DelayedMatrixStats::colVars(DelayedArray(exprs_mat))
  exprs_mat <- sweep(exprs_mat, MARGIN=2, # We scale by the standard deviation, which also changes the centre.
             STATS=sqrt(vars), FUN="/", 
             check.margin=FALSE) # divide by the square root of vars

  ## PCA
  pca <- prcomp(exprs_mat)

  emt_w <- pca$rotation[,1]
  emt_w <- emt_w[order(emt_w, decreasing = T)]
  
  emt_w 
}
```

```{r select-markers-by-pcs}
## genes with too good correlations can go (i.e. cor > 0.8)
# cor(t(log1p(exprs_tcga)[new_markers$EMT,]))
new_markers <- list()

# new_markers$progenitor <- markers$progenitor
out <- pca_emt(markers = markers$C3)
new_markers$C3 <- names(out)[abs(out) > 0.2]

out <- pca_emt(markers = markers$C4)
new_markers$C4 <- names(out)[abs(out) > 0.2]

out <- pca_emt(markers = markers$EMT)
new_markers$EMT <- names(out)[abs(out) > 0.2]
# cor(t(log1p(exprs_tcga)[new_markers$EMT,]))

out <- pca_emt(markers = markers$C10)
new_markers$C10 <- names(out)[abs(out) > 0.2]

out <- pca_emt(markers = markers$ciliated)

new_markers$ciliated <- names(out)[abs(out) > 0.2]
new_markers$ciliated <- new_markers$ciliated[c(1,2,3,8,9,10,13,14)] # remove redundant genes by cor > 0.8
# str(new_markers)
```

```{r length-of-signaure}
length(unlist(new_markers))
```

There are 52 genes in total

```{r count-number-of-genes-in-each-signature}
lapply(new_markers, length)
```


# BSEQ-sc: compute the reference matrix

```{r bseq-signature-matrix}
sig_matrix_new <- bseq_my(markers = new_markers)
## upper limit
quantile(sig_matrix_new, 0.99)

sig_matrix_new[sig_matrix_new > quantile(sig_matrix_new, 0.99)] <- quantile(sig_matrix_new, 0.99) # average the too high value
```

```{r save-plot-and-data-about-signature-matrix, include=F}
# pdf("plots_explore/20190207/sig_matrix_new.pdf", width = 4, height = 4)
# plotBasis(sig_matrix_new, new_markers, Colv = NA, Rowv = NA, layout = '_', col = 'Blues')
# dev.off()

# saveRDS(sig_matrix_new, "../../scFT-paper_rds/20190213sig_matrix_new.rds", compress = T)
# saveRDS(new_markers, "../../scFT-paper_rds/20190213new_markers.rds", compress = T)
# saveRDS(fresheset, "rds/20190213fresheset.rds", compress = T)

# write.csv(sig_matrix_new,"20190213_selected_signature_matrix_52genes.csv")
```

```{r}
gc()
rm(secretory)
rm(fresh)
rm(fresheset)
```

# Deconvolution and survival analysis

## Cibersort TCGA

Use the reference matrix to decompose TCGA data

```{r deconvolute-TCGA-data-and-heatmap}
svr_fit_new <- cibersort_my(sig_matrix = sig_matrix_new, exprs_unlogged = exprs_tcga)
eset$EMT <- svr_fit_new[,"EMT"]
head(svr_fit_new)

# write.csv(svr_fit_new,"../../scFT-paper_rds/20191125TCGA_deconvolutionResults.csv")
```


### Barplots

Barplots

```{r tcga-stacked-barplot-deconvolution-results}
svr_fit_new_diff <- svr_fit_new[svr_fit_new[,1] > 0.5,]
svr_fit_new_diff <- svr_fit_new_diff[order(svr_fit_new_diff[,1], decreasing = F),]
svr_fit_new_krt <- svr_fit_new[svr_fit_new[,2] > 0.5,]
svr_fit_new_krt <- svr_fit_new_krt[order(svr_fit_new_krt[,2], decreasing = F),]
svr_fit_new_emt <- svr_fit_new[svr_fit_new[,3] > 0.5,]
svr_fit_new_emt <- svr_fit_new_emt[order(svr_fit_new_emt[,3], decreasing = F),]
svr_fit_new_cc <- svr_fit_new[svr_fit_new[,4] > 0.5,]
svr_fit_new_cc <- svr_fit_new_cc[order(svr_fit_new_cc[,4], decreasing = F),]
svr_fit_new_others <- svr_fit_new[!rownames(svr_fit_new) %in% c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc)),]

plot.data <- data.frame(sample = rep(rownames(svr_fit_new), 5),
                        state = rep(c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), each = nrow(svr_fit_new)),
                        score = c(svr_fit_new[,1],
                                  svr_fit_new[,2],
                                  svr_fit_new[,3],
                                  svr_fit_new[,4],
                                  svr_fit_new[,5]))
plot.data$sample <- factor(plot.data$sample, levels = c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                        rownames(svr_fit_new_others)))
plot.data$state <- factor(plot.data$state, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"))
# differentiated #A16BA3
# F6A000 KRT7
# FFDE00 EMT
# CC #4DD9FF
#F3A1B6 Ciliated
```

```{r  tcga-stacked-barplot-plotting, fig.width=10, fig.height=3}
ggplot(plot.data) + geom_bar(aes(y = score, x = sample, fill = state),
                           stat="identity") + theme_pubclean() + 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6"), 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  xlab("TCGA bulk tumour samples") + ylab("Proportion") +
  theme(axis.ticks.x  = element_blank(), axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size= 15),
        legend.text = element_text(size = 12),
        legend.title  = element_blank())

# ggsave("plots/Fig4_TCGA_stackedBarplot_20190509.png", width = 10, height = 5)
```

```{r tcga-deconvolutin-barplot}
plot.data2 <- data.frame(State = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"),
                         Fraction = colSums(svr_fit_new[,1:5] > 0)/nrow(svr_fit_new))
plot.data2$State <- factor(plot.data2$State, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")[5:1])
ggplot(plot.data2) + geom_bar(aes(y = Fraction,  x = State , fill = State),stat = "identity")+ theme_pubclean() + coord_flip()+ 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6")[5:1], 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  theme(legend.position = "none", axis.text = element_text(size = 16), axis.title = element_text(size = 20))
# ggsave("plots/Fig4_TCGA_barplot_20190509.png", width = 4, height = 5)
```


### t-SNE plot

```{r TCGA-deconvolution-tSNE-plot, fig.width=12, fig.height=2.4}
colnames(svr_fit_new)[c(1,2,4,5)] <- c("Differentiated","KRT17","Cell_cycle","Ciliated")

set.seed(158163)
tsne_res <- tsne::tsne(svr_fit_new[,1:5], k = 2, perplexity = round(0.05*nrow(svr_fit_new)))

df_tsne <- data.frame(TSNE_1 = tsne_res[,1],
                     TSNE_2 = tsne_res[,2],
                     EMT = svr_fit_new[,"EMT"],
                     Cell_cycle = svr_fit_new[,"Cell_cycle"],
                     Differentiated = svr_fit_new[,1],
                     KRT17 = svr_fit_new[,2],
                     Ciliated =  svr_fit_new[,5],
                     sample = rownames(svr_fit_new))
p1 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = EMT)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p2 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Cell_cycle)) + geom_point(size =1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p3 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Differentiated)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p4 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = KRT17)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p5 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Ciliated)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
cowplot::plot_grid(p1,p2,p3,p4,p5, ncol = 5)
# ggsave("plots/Fig6_TSNE_deconvolution_TCGA.pdf",device = "pdf", width = 12.5, height = 3)
```

```{r tcga-deconvolution-regular-heatmap, include=F}
df_tsne$Type <- "Mixture"
df_tsne$Type[df_tsne$EMT > 0.5] <- "EMT"
df_tsne$Type[df_tsne$Cell_cycle > 0.5] <- "Cell cycle"
df_tsne$Type[df_tsne$Differentiated > 0.5] <- "Differentiated"
df_tsne$Type[df_tsne$KRT17 > 0.5] <- "KRT17 Cluster"
df_tsne$Type[df_tsne$Ciliated > 0.5] <- "Ciliated"
# RColorBrewer::display.brewer.all()
RColorBrewer::brewer.pal(8, "Set1")
# "#E41A1C" "#377EB8" "#4DAF4A" "#984EA3" "#FF7F00" "#FFFF33" "#A65628" "#F781BF"
my_colour <- RColorBrewer::brewer.pal(8, "Set1")[c(2,4,6,7)]
# ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Type)) + geom_point() + theme_classic() + scale_color_manual(values=c(my_colour,"grey"))
# ggsave("plots_explore/20190423PCA_deconvolution/TSNE_deconvolution_TCGA2.png", width = 5, height = 3.5)

plot.htmp <- pheatmap::pheatmap(svr_fit_new[,1:5], cluster_cols = F, show_rownames = F, border_color = NA, cellwidth = 20, cellheight = 1)
# gheat <- ggplotify::as.ggplot(plot.htmp )
# ggsave(gheat, filename = "plots/Fig4_TCGA_heatmap.pdf", device = "pdf", width = 8, height = 10, units = "in")

```

### Survival analysis

Univariate survival anlaysis gave significant p-value and hazard ratio ~ 2.3

```{r survival-analysis-univariate-tcga}
# eset$emt_new <- scale(svr_fit_new[,"EMT"], scale = T, center = T)
eset$emt_new <- svr_fit_new[,"EMT"]
fit <- coxph(Surv(OS.time, OS)~emt_new, data=pData(eset))
summary(fit)
```

Multivariate analysis

```{r survival-analysis-multivariate-tcga}
fit <- coxph(Surv(OS.time, OS)~emt_new+stage_dichotomized, data=pData(eset))
summary(fit)
```


## Cibersort AOCS

```{r aocs-deconvolution-and-survival-analysis}
svr_fit_tot <- cibersort_my(sig_matrix = sig_matrix_new, exprs_unlogged = exprs_tot)
toteset$emt.new <- svr_fit_tot[,"EMT"]
# saveRDS(svr_fit_tot, "results/20190502Deconvolution_Tothill.rds", compress = T)
# only for high-grade and late stage patients
fit_tot <- coxph(Surv(time.to.death, OS)~emt.new, data=pData(toteset)[toteset$histological.subtype2 == "ser" ,])
summary(fit_tot)
```

The univariate survival analysis on AOCS data (serous OvCa only) showed HR = 2.53 with p = 0.000882.

Multivariate analysis

```{r}
fit_tot <- coxph(Surv(time.to.death, OS)~emt.new+summarystage+summarygrade, data=pData(toteset)[toteset$histological.subtype2 == "ser" ,])
summary(fit_tot)
```


### Barplots

```{r aocs-prepare-stacked-barplots}
svr_fit_new_diff <- svr_fit_tot[svr_fit_tot[,1] > 0.5,]
svr_fit_new_diff <- svr_fit_new_diff[order(svr_fit_new_diff[,1], decreasing = F),]
svr_fit_new_krt <- svr_fit_tot[svr_fit_tot[,2] > 0.5,]
svr_fit_new_krt <- svr_fit_new_krt[order(svr_fit_new_krt[,2], decreasing = F),]
svr_fit_new_emt <- svr_fit_tot[svr_fit_tot[,3] > 0.5,]
svr_fit_new_emt <- svr_fit_new_emt[order(svr_fit_new_emt[,3], decreasing = F),]
svr_fit_new_cc <- svr_fit_tot[svr_fit_tot[,4] > 0.5,]
svr_fit_new_cc <- svr_fit_new_cc[order(svr_fit_new_cc[,4], decreasing = F),]
svr_fit_new_cil <- svr_fit_tot[svr_fit_tot[,5] > 0.5,]
svr_fit_new_cil <- svr_fit_new_cil[order(svr_fit_new_cil[,5], decreasing = F),]
svr_fit_new_others <- svr_fit_tot[!rownames(svr_fit_tot) %in% c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                               rownames(svr_fit_new_cil)),]

plot.data <- data.frame(sample = rep(rownames(svr_fit_tot), 5),
                        state = rep(c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), each = nrow(svr_fit_tot)),
                        score = c(svr_fit_tot[,1],
                                  svr_fit_tot[,2],
                                  svr_fit_tot[,3],
                                  svr_fit_tot[,4],
                                  svr_fit_tot[,5]))
plot.data$sample <- factor(plot.data$sample, levels = c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                        rownames(svr_fit_new_cil),
                                                        rownames(svr_fit_new_others)))
plot.data$state <- factor(plot.data$state, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"))
# differentiated #A16BA3
# F6A000 KRT7
# FFDE00 EMT
# CC #4DD9FF
#F3A1B6 Ciliated
```

```{r aocs-deconvolution-stacked-barplot-plotting, fig.width=10, fig.height=3}
ggplot(plot.data) + geom_bar(aes(y = score, x = sample, fill = state),
                           stat="identity") + theme_pubclean() + 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6"), 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  xlab("AOCS bulk tumour samples") + ylab("Proportion") +
  theme(axis.ticks.x  = element_blank(), axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size= 15),
        legend.text = element_text(size = 12),
        legend.title  = element_blank())

# ggsave("plots/Fig4_AOCS_stackedBarplot_20190509.png", width = 10, height = 5)
```

```{r aocs-deconvoltuion-barplot}
plot.data2 <- data.frame(State = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"),
                         Fraction = colSums(svr_fit_tot[,1:5] > 0)/nrow(svr_fit_tot))
plot.data2$State <- factor(plot.data2$State, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")[5:1])
ggplot(plot.data2) + geom_bar(aes(y = Fraction,  x = State , fill = State),stat = "identity")+ theme_pubclean() + coord_flip()+ 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6")[5:1], 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  theme(legend.position = "none", axis.text = element_text(size = 16), axis.title = element_text(size = 20))
# ggsave("plots/Fig4_AOCS_barplot_20190509.png", width = 4, height = 5)
```


### Heatmap: TCGA & AOCS

We plot the heatmap for the diagram

```{r tcga-deconvolution-heatmap-for-diagram}
data.heatmap <- t(svr_fit_new[,1:5])
pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
  show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)

# pdf("plots/Fig4_heatmap_mini_tcga_20190510.pdf", height = 5, width = 5)
# pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
#   show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)
# dev.off()
```

```{r aocs-deconvolution-heatmap-for-diagram}
data.heatmap <- t(svr_fit_tot[,1:5])
pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
  show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)

# pdf("plots/Fig4_heatmap_mini_aocs_20190510.pdf", height = 5, width = 5)
# pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
#   show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)
# dev.off()
```

## Cibersort OXO-PCR

```{r cibersort-oxopcr}
svr_fit_oxo <- cibersort_my(sig_matrix = sig_matrix_new, exprs_unlogged = cpm(as.matrix(oxopcr_pre)))
```

### Heatmap


```{r oxopcr_heatmap}
rownames(svr_fit_oxo) <- gsub(rownames(svr_fit_oxo), pattern = "X", replacement = "")
colnames(svr_fit_oxo)[1:5] <- c("Differentiated", "KRT17","EMT","Cell cycle","Ciliated")
rownames(svr_fit_oxo)[rownames(svr_fit_oxo) == "1G3_new"] <- "1016_1G3new"
```

```{r oxopcr_heatmap-plotting, include = F}
plot.htmp <- pheatmap::pheatmap(svr_fit_oxo[order(oxopcr_info_pre$`PATIENT ID`),1:5], cluster_cols = F, cluster_rows = F,show_rownames = T, border_color = NA, cellwidth = 20, cellheight = 8, fontsize = 8)
gheat <- ggplotify::as.ggplot(plot.htmp)
# ggsave(gheat, filename = "plots/Fig4_OXOPCR_heatmap20190316.pdf", device = "pdf", width = 8, height = 12, units = "in")
```

### Stacked barplot

```{r oxopcr-deconvolution-stacked-barplot}
svr_fit_new_diff <- svr_fit_oxo[svr_fit_oxo[,1] > 0.5,]
svr_fit_new_diff <- svr_fit_new_diff[order(svr_fit_new_diff[,1], decreasing = F),]
svr_fit_new_krt <- svr_fit_oxo[svr_fit_oxo[,2] > 0.5,]
svr_fit_new_krt <- svr_fit_new_krt[order(svr_fit_new_krt[,2], decreasing = F),]
svr_fit_new_emt <- svr_fit_oxo[svr_fit_oxo[,3] > 0.5,]
svr_fit_new_emt <- svr_fit_new_emt[order(svr_fit_new_emt[,3], decreasing = F),]
svr_fit_new_cc <- svr_fit_oxo[svr_fit_oxo[,4] > 0.5,]
svr_fit_new_cc <- svr_fit_new_cc[order(svr_fit_new_cc[,4], decreasing = F),]
svr_fit_new_others <- svr_fit_oxo[!rownames(svr_fit_oxo) %in% c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc)),]

plot.data <- data.frame(sample = rep(rownames(svr_fit_oxo), 5),
                        state = rep(c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                                    each = nrow(svr_fit_oxo)),
                        score = c(svr_fit_oxo[,1],
                                  svr_fit_oxo[,2],
                                  svr_fit_oxo[,3],
                                  svr_fit_oxo[,4],
                                  svr_fit_oxo[,5]))
plot.data$sample <- factor(plot.data$sample, levels = c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                        rownames(svr_fit_new_others)))
plot.data$state <- factor(plot.data$state, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"))
# differentiated #A16BA3
# F6A000 KRT7
# FFDE00 EMT
# CC #4DD9FF
#F3A1B6 Ciliated
```

```{r oxopcr-deconvolution-stacked-barplot-plotting, fig.width=10, fig.height=3}
ggplot(plot.data) + geom_bar(aes(y = score, x = sample, fill = state),
                           stat="identity") + theme_pubclean() + 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6"), 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  xlab("TCGA bulk tumour samples") + ylab("Proportion") +
  theme(axis.ticks.x  = element_blank(), axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size= 15),
        legend.text = element_text(size = 12),
        legend.title  = element_blank())

# ggsave("plots/Fig5_OXOPCR_stackedBarplot_20190509.png", width = 7, height = 3)
```

## Permutation test

In the permutation test, we select 90 percent of the samples from each dataset and reran the survival analysis to make sure that the association between EMT scores and the overall survival is robust (not affected by few extreme cases).

For TCGA data

```{r permutation-tcga}
Perm <- 500
prob <- 0.9

p_val_tcga <- c()

for(itor in 1:Perm){
  idx <- sample(x = 1:ncol(eset), size = round(ncol(eset) * prob), replace = F)
  fit <- coxph(Surv(OS.time, OS)~emt_new, data=pData(eset)[idx,])
  p_val_tcga <- c(p_val_tcga, summary(fit)$coef[,5])
}

sum(p_val_tcga <= 0.05)/500 * 100
```

For the high-grade tumours in Tothill's data

```{r permutation-aocs}
Perm <- 500
prob <- 0.9 # select 90% samples each time

p_val_tot <- c()

for(itor in 1:Perm){
  idx <- sample(x = which(toteset$Grade >= 2 & toteset$histological.subtype2 == "ser"), size = round(sum(toteset$Grade >= 2 & toteset$histological.subtype2 == "ser", na.rm = TRUE) * prob), replace = F)
  fit <- coxph(Surv(time.to.death, OS)~emt.new, data=pData(toteset)[idx,])
  p_val_tot <- c(p_val_tot, summary(fit)$coef[,5])
}

sum(p_val_tot <= 0.05)/500 * 100
```


## CuratedOvarianCancer database

```{r curatedovariancancer-deconvolution}
## read in the list of selected CuratedOvarianCancer database
esets <- readRDS("../../scFT-paper_rds/20190213CuratedOvarianData_esets.rds")

svr_fit <- list() # record cibersort results
fit <- list() # survival analysis results
esets_tumor <- list() 

# go through all the datasets, run cibersort
for(itor in 1:length(esets)) { 
 
  ## tumor data/remove the benign or borderline samples
  eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
  
  ## deconvolution
  if(max(na.omit(eset_tmp@assayData$exprs)) < 50) { ## if log-space, unlog it
    svr_fit[[itor]] <- cibersort_my(sig_matrix = sig_matrix_new, 
                                    exprs_unlogged = exp(na.omit(eset_tmp@assayData$exprs)))
  } else {         # non-log space
    svr_fit[[itor]] <- cibersort_my(sig_matrix = sig_matrix_new, 
                                    exprs_unlogged = eset_tmp@assayData$exprs)
  }
}
```



### Manually iterations

We applied survival analysis on the serous ovarian cancer cases with stage and grade as covariate.

```{r}
itor=1
eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
# # survival data
eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

table(eset_tmp$sample_type)
table(eset_tmp$histological_type)
table(eset_tmp$summarygrade)
table(eset_tmp$summarystage)
eset_tmp <- eset_tmp[,eset_tmp$summarystage == "late"]
fit[[itor]] <- coxph(y~emt, eset_tmp) 
esets_tumor[[itor]] <- eset_tmp

#####

itor=2
eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
# # survival data
eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

table(eset_tmp$sample_type)
table(eset_tmp$histological_type)
table(eset_tmp$summarygrade)
table(eset_tmp$summarystage)

fit[[itor]] <- coxph(y~emt+summarygrade, eset_tmp) 
esets_tumor[[itor]] <- eset_tmp

#####

itor=3
eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
# # survival data
eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

table(eset_tmp$sample_type)
table(eset_tmp$histological_type)
eset_tmp <- eset_tmp[,eset_tmp$histological_type=="ser"]
table(eset_tmp$summarygrade)
table(eset_tmp$summarystage)
# early  late 
#    17    62 


fit[[itor]] <- coxph(y~emt+summarygrade+summarystage, eset_tmp) 
esets_tumor[[itor]] <- eset_tmp


#####

itor=4
eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
# # survival data
eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

table(eset_tmp$sample_type)
table(eset_tmp$histological_type)
# eset_tmp <- eset_tmp[,eset_tmp$histological_type=="ser"]
table(eset_tmp$summarygrade)
table(eset_tmp$summarystage)
# late 
#  185
fit[[itor]] <- coxph(y~emt, eset_tmp) 
esets_tumor[[itor]] <- eset_tmp

#####

itor=5
eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
# # survival data
eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

table(eset_tmp$sample_type)
table(eset_tmp$histological_type)
# eset_tmp <- eset_tmp[,eset_tmp$histological_type=="ser"]
table(eset_tmp$summarygrade)
table(eset_tmp$summarystage)
# late 
#  260 


fit[[itor]] <- coxph(y~emt+summarygrade, eset_tmp) 
esets_tumor[[itor]] <- eset_tmp

#####

itor=6
eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
# # survival data
eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

table(eset_tmp$sample_type)
table(eset_tmp$histological_type)
eset_tmp <- eset_tmp[,eset_tmp$histological_type=="ser"]
table(eset_tmp$summarygrade)
table(eset_tmp$summarystage)
# early  late 
#     5   166 
fit[[itor]] <- coxph(y~emt+summarygrade+ summarystage, eset_tmp) 
esets_tumor[[itor]] <- eset_tmp

#####

itor=7
eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
# # survival data
eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

eset_tmp <- eset_tmp[,!is.na(eset_tmp$days_to_death)]

table(eset_tmp$sample_type)
table(eset_tmp$histological_type)
eset_tmp <- eset_tmp[,eset_tmp$histological_type=="ser"]
table(eset_tmp$summarygrade)
table(eset_tmp$summarystage)
# early  late 
#     5   108 
fit[[itor]] <- coxph(y~emt+summarygrade+summarystage, eset_tmp) 
esets_tumor[[itor]] <- eset_tmp
```


### Generate table 1

```{r organise-table}
df <- matrix(NA, nrow = 7, ncol = 6)
colnames(df) <- c("nsamples","nevent","hazard_ratio","pvalue","lower.CI95","upper.CI95")
df <- as.data.frame(df)
for(itor in 1:7){
  df$lower.CI95[itor] <- summary(fit[[itor]])$conf.int[1,3]
  df$upper.CI95[itor] <- summary(fit[[itor]])$conf.int[1,4]
  df$hazard_ratio[itor] <- summary(fit[[itor]])$conf.int[1,1]
  df$pvalue[itor] <- summary(fit[[itor]])$coefficient[1,5]
  df$nsamples[itor] <- summary(fit[[itor]])$n
  df$nevent[itor] <- summary(fit[[itor]])$nevent
}

df$database <- c("E.MTAB.386", "GSE13876", "GSE26193",
              "GSE26712", "GSE32062.GPL6480",
              "GSE49997", "GSE51088")
  
# write.csv(df, "../revision_tables/20191009CuratedOvarianData_results_multivariable_test.csv", row.names = F)

## IC95 error bar plot----
df <- df[order(df$hazard_ratio, decreasing = T),]
df <- df[!grepl("TCGA",df$database),]
ggplot(df[which(df$nsamples > 100),], aes(y = database, x = hazard_ratio))+ geom_point(shape=21, fill = "black",size=4) +
  geom_errorbarh(aes(xmax = upper.CI95, xmin = lower.CI95), height=0.2, size=1) + theme_survminer() +xlim(c(0, 10)) +
  geom_vline(xintercept = 1, lty = 3, col = "black") + xlab("Hazard ratio") +ylab("")

# ggsave("plots_explore/20190213Hazard_ratio_CI95_CuratedOvarianData.png",width = 5, height = 6)
```

```{r print-CuratedOvarianData-survival-results}
knitr::kable(df)
```


# Put all datasets together

```{r merge-all-data}
## put all datasets together
esets_tumor[[6]] <- esets_tumor[[6]][,!is.na(esets_tumor[[6]]$alt_sample_name)]
df_all <- data.frame(database = c(rep("E.MTAB.386",ncol(esets_tumor[[1]])),
                                  rep("GSE13876",ncol(esets_tumor[[2]])),
                                  rep("GSE26193",ncol(esets_tumor[[3]])),
                                  rep("GSE26712",ncol(esets_tumor[[4]])),
                                  rep("GSE32062.GPL6480",ncol(esets_tumor[[5]])),
                                  rep("GSE49997",ncol(esets_tumor[[6]])),
                                  rep("GSE51088",ncol(esets_tumor[[7]])),
                                  rep("AOCS", ncol(toteset)),
                                  rep("TCGA", ncol(eset))),
                     
                     sample_type = c(esets_tumor[[1]]$sample_type,esets_tumor[[2]]$sample_type,
                                     esets_tumor[[3]]$sample_type,esets_tumor[[4]]$sample_type,
                                     esets_tumor[[5]]$sample_type,esets_tumor[[6]]$sample_type,
                                     esets_tumor[[7]]$sample_type, rep("tumor",ncol(toteset)),
                                     rep("tumor", ncol(eset))),
                     
                     grade = c(esets_tumor[[1]]$summarygrade,esets_tumor[[2]]$summarygrade,
                               esets_tumor[[3]]$summarygrade,esets_tumor[[4]]$summarygrade,
                              esets_tumor[[5]]$summarygrade,
                               esets_tumor[[6]]$summarygrade,esets_tumor[[7]]$summarygrade,
                               toteset$summarygrade, 
                               rep("high", ncol(eset))),
                     stage = c(esets_tumor[[1]]$summarystage,esets_tumor[[2]]$summarystage,
                               esets_tumor[[3]]$summarystage,esets_tumor[[4]]$summarystage,
                              esets_tumor[[5]]$summarystage,
                               esets_tumor[[6]]$summarystage,esets_tumor[[7]]$summarystage,
                               toteset$summarystage,
                               eset$stage_dichotomized),
                     histological_type = c(esets_tumor[[1]]$histological_type,esets_tumor[[2]]$histological_type,
                                           esets_tumor[[3]]$histological_type,esets_tumor[[4]]$histological_type,
                                        esets_tumor[[5]]$histological_type,esets_tumor[[6]]$histological_type,
                                        esets_tumor[[7]]$histological_type, tolower(toteset$histological.subtype),
                                        rep("ser", ncol(eset)))
                     )

toteset$vital_status <- NA
toteset$vital_status[toteset$patient.status != "D"] <- "living"
toteset$vital_status[toteset$patient.status == "D"] <- "deceased"

df_all$OS <- c(esets_tumor[[1]]$vital_status,esets_tumor[[2]]$vital_status,
               esets_tumor[[3]]$vital_status,esets_tumor[[4]]$vital_status,
               esets_tumor[[5]]$vital_status,esets_tumor[[6]]$vital_status,
               esets_tumor[[7]]$vital_status,toteset$vital_status, tolower(eset$vital_status)
               )

df_all$days_to_death <- c(esets_tumor[[1]]$days_to_death,esets_tumor[[2]]$days_to_death,
                          esets_tumor[[3]]$days_to_death,esets_tumor[[4]]$days_to_death,
                          esets_tumor[[5]]$days_to_death,esets_tumor[[6]]$days_to_death,
                          esets_tumor[[7]]$days_to_death,
                          toteset$time.to.death*30,
                          eset$OS.time)

# df_all <- df_all[df_all$sample_type %in% c("metastatic","tumor"),]

df_all$EMT <- c( esets_tumor[[1]]$emt, esets_tumor[[2]]$emt,
                  esets_tumor[[3]]$emt,  esets_tumor[[4]]$emt,
                  esets_tumor[[5]]$emt, esets_tumor[[6]]$emt,
                  esets_tumor[[7]]$emt,
                 toteset$emt.new, eset$EMT)


# saveRDS(fit,"results/20181122CuratedOvarianData_survival_list.rds",compress = T)
df_all$years_to_death <- df_all$days_to_death/365
df_all$y <- Surv(df_all$years_to_death,
                     df_all$OS == "deceased")

df_all <- df_all[which(df_all$histological_type == "ser"),]

# survival analysis on the merged data
fit <- coxph(y~EMT+grade+stage, data = df_all) 
summary(fit)
```

## Overall KM plot

```{r dichotomise-EMTgroups}
df_all$EMT_group <- NA

for(itor in 1:9){
   emt_tmp <- median(df_all$EMT[df_all$database == unique(df_all$database)[itor]])
   df_all$EMT_group[df_all$database == unique(df_all$database)[itor] & df_all$EMT > emt_tmp] <- "high"
   df_all$EMT_group[df_all$database == unique(df_all$database)[itor] & df_all$EMT <= emt_tmp] <- "low"
}

table(df_all$EMT_group, df_all$database)
```

```{r km-curves-merged-dataset}
fit1 <- survfit(y~EMT_group, data = df_all)
ggsurvplot(fit1, data = df_all, pval = F) + xlab("Years") 

# ggsave("../revision_plots/main/20191009_survival_pooled_nine.png", width = 5, height = 4)
```

# AOCS: Grade & ciliated cells

```{r aocs-ciliated}
toteset$ciliated.new <- svr_fit_tot[,"ciliated"]
ggplot(as.data.frame(pData(toteset))[!is.na(toteset$grade),], aes(x = as.factor(grade), y = ciliated.new)) + geom_violin(scale = "width") + theme_classic() + geom_quasirandom(alpha = 0.4, width = 0.3) +
  xlab("Grades")+ ylab("Score of ciliated signature")

# ggsave("plots/20190215AOCS_grade_ciliated.pdf", width = 4, height = 3)
```

```{r aocs-ciliated-wilcox}
wilcox.test(toteset$ciliated.new[toteset$grade == 1], toteset$ciliated.new[toteset$grade > 1], alternative = "greater")
```


# Save data

```{r sav-tcga-data}
# saveRDS(eset, "../../scFT-paper_rds/20191009_esetTCGA_afterDeconvolution.rds")
# write.csv(sig_matrix_new, "../../scFT-paper_rds/sig_matrix_new20200109.csv")
```

# Technical

```{r sessionInfo}
sessionInfo()

# R version 3.5.2 (2018-12-20)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Mojave 10.14.4
# 
# Matrix products: default
# BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
# [1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] RColorBrewer_1.1-2          edgeR_3.24.3                limma_3.38.3                viridis_0.5.1               viridisLite_0.3.0          
#  [6] ggrepel_0.8.0               ggbeeswarm_0.6.0            scater_1.10.1               SingleCellExperiment_1.4.1  SummarizedExperiment_1.12.0
# [11] DelayedArray_0.8.0          BiocParallel_1.16.5         matrixStats_0.54.0          GenomicRanges_1.34.0        GenomeInfoDb_1.18.1        
# [16] powerSurvEpi_0.1.0          survival_2.43-3             survminer_0.4.3             ggpubr_0.2                  magrittr_1.5               
# [21] ggplot2_3.2.0.9000          preprocessCore_1.44.0       e1071_1.7-0.1               xbioc_0.1.15                AnnotationDbi_1.44.0       
# [26] IRanges_2.16.0              S4Vectors_0.20.1            MASS_7.3-51.1               bseqsc_1.0                  Biobase_2.42.0             
# [31] BiocGenerics_0.28.0         csSAM_1.4                   Rcpp_1.0.0                 
# 
# loaded via a namespace (and not attached):
#   [1] colorspace_1.4-0         class_7.3-15             modeltools_0.2-22        mclust_5.4.2             XVector_0.22.0          
#   [6] rstudioapi_0.9.0         flexmix_2.3-14           bit64_0.9-7              mvtnorm_1.0-8            codetools_0.2-16        
#  [11] splines_3.5.2            doParallel_1.0.14        robustbase_0.93-3        knitr_1.21               Formula_1.2-3           
#  [16] broom_0.5.1              gridBase_0.4-7           km.ci_0.5-2              cluster_2.0.7-1          kernlab_0.9-27          
#  [21] pheatmap_1.0.12          HDF5Array_1.10.1         compiler_3.5.2           rvcheck_0.1.3            backports_1.1.3         
#  [26] assertthat_0.2.0         Matrix_1.2-15            lazyeval_0.2.1           htmltools_0.3.6          tools_3.5.2             
#  [31] bindrcpp_0.2.2           gtable_0.2.0             glue_1.3.0               GenomeInfoDbData_1.2.0   reshape2_1.4.3          
#  [36] dplyr_0.7.8              NMF_0.22                 trimcluster_0.1-2.1      nlme_3.1-137             iterators_1.0.10        
#  [41] fpc_2.1-11.1             DelayedMatrixStats_1.4.0 xfun_0.4                 stringr_1.4.0            rngtools_1.3.1          
#  [46] dendextend_1.9.0         DEoptimR_1.0-8           zlibbioc_1.28.0          zoo_1.8-4                scales_1.0.0            
#  [51] BiocInstaller_1.32.1     rhdf5_2.26.2             yaml_2.2.0               memoise_1.1.0            gridExtra_2.3           
#  [56] KMsurv_0.1-5             pkgmaker_0.27            stringi_1.2.4            RSQLite_2.1.1            foreach_1.4.4           
#  [61] bibtex_0.4.2             rlang_0.4.0              pkgconfig_2.0.2          prabclus_2.2-7           bitops_1.0-6            
#  [66] evaluate_0.13            lattice_0.20-38          Rhdf5lib_1.4.2           purrr_0.3.0              bindr_0.1.1             
#  [71] labeling_0.3             cmprsk_2.2-7             cowplot_0.9.4            bit_1.1-14               tidyselect_0.2.5        
#  [76] plyr_1.8.4               R6_2.3.0                 generics_0.0.2           DBI_1.0.0                pillar_1.3.1            
#  [81] whisker_0.3-2            withr_2.1.2              RCurl_1.95-4.11          nnet_7.3-12              tsne_0.1-3              
#  [86] tibble_2.0.1             crayon_1.3.4             survMisc_0.5.5           rmarkdown_1.11           locfit_1.5-9.1          
#  [91] grid_3.5.2               data.table_1.12.0        blob_1.1.1               digest_0.6.18            diptest_0.75-7          
#  [96] xtable_1.8-3             tidyr_0.8.2              gridGraphics_0.3-0       munsell_0.5.0            ggplotify_0.0.3         
# [101] beeswarm_0.2.3           registry_0.5             vipor_0.4.5             
```
