---
title: "Visulisation of secretory clusters"
author: "Zhiyuan Hu"
date: "12/11/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, echo = FALSE)
library(SingleCellExperiment) #1.4.1
library(scater)#1.10.1
library(edgeR) #3.24.3
library(limma) #3.38.3
library(dplyr) #0.7.8
library(scales)
library(reshape2)
```


```{r}
secretory <- readRDS("../../scFT-paper_rds/20190120Fresh_secretory_9clusters_clincluster.rds")
markers2 <- read.csv("../tables/20190120Clincluster_fresh_secretory_9clusters_markers.csv", as.is = T)
```

## Visualisation

#### TSNE

```{r}
secretory <- scran::computeSumFactors(secretory)
secretory <- scater::normalize(secretory)
```


```{r}
set.seed(123456)
secretory <- runTSNE(object = secretory, ncomponents = 2, 
                     feature_set = rownames(secretory)[rowData(secretory)$high.var],
                  exprs_values = "normcounts",
                  perplexity = min(50, floor(ncol(secretory)/5)))
# secretory <- runTSNE(object = secretory, ncomponents = 2, 
#                      feature_set = markers2$gene[markers2$cluster %in% c("C10","C3","C4","C8")],
#                      exprs_values = "logcounts", 
#                      perplexity = min(50, floor(ncol(secretory)/5)))
secretory <- scater::runUMAP(object = secretory, ncomponents = 2, use_dimred = "PCA", n_dimred = 15,exprs_values = "normcounts")
```

```{r}
table(secretory$clincluster_final)
```

```{r}
secretory$cell_type <- NA
secretory$cell_type[secretory$clincluster_final == "C10"] <- "Cell cycle (C9)"
secretory$cell_type[secretory$clincluster_final == "C3"] <- "Differentiated (C3)"
secretory$cell_type[secretory$clincluster_final == "C4"] <- "KRT17 (C4)"
secretory$cell_type[secretory$clincluster_final == "C6"] <- "Stress (C6)"
secretory$cell_type[secretory$clincluster_final == "C8"] <- "EMT (C7)"
secretory$cell_type[secretory$clincluster_final == "C9"] <- "Immune (C8)"

secretory$Cell_subtype <- secretory$cell_type

```


```{r}
plotTSNE(secretory[,secretory$clincluster_final %in% c("C3","C4","C8","C10","C9")], colour_by = "Cell_subtype", shape_by = "Patient2") +
    # theme(legend.position = "top" )  + 
    xlab("TSNE_1") + ylab("TSNE_2")
# ggsave("plots/20190508_TSNE_secretory.png", width=5, height = 3.5)
```

```{r}
p1 <- plotTSNE(secretory[,secretory$clincluster_final %in% c("C3","C4","C8","C10","C9")], colour_by = "Cell_subtype") +
    # theme(legend.position = "top" )  + 
    xlab("TSNE_1") + ylab("TSNE_2")

p2 <- plotTSNE(secretory[,secretory$clincluster_final %in% c("C3","C4","C8","C10","C9")], colour_by = "Patient2") +
    # theme(legend.position = "top" )  + 
    xlab("TSNE_1") + ylab("TSNE_2")
cowplot::plot_grid(p1, p2, rel_widths = c(1.1, 1 ))
# ggsave("plots/20190508_TSNE_secretory.png", width=5, height = 3.5)
```

```{r}
plotUMAP(secretory[,secretory$clincluster_final %in% c("C3","C4","C8","C10","C9")], colour_by = "Cell_subtype") +
    # theme(legend.position = "top" )  + 
    xlab("UMAP_1") + ylab("UMAP_2")
# ggsave("../../revision_analysis_20190827/revision_plots/SI/secretory_UMAP_colSubtype20191007.png", width=5, height = 3.5)
```

```{r}
plotTSNE(secretory[,secretory$clincluster_final %in% c("C3","C4","C8","C10","C9")], colour_by = "Patient2") +
    # theme(legend.position = "top" )  + 
    xlab("TSNE_1") + ylab("TSNE_2")
# ggsave("../../revision_analysis_20190827/revision_plots/SI/secretory_TSNE_colPatient20191007.png", width=5, height = 3.5)
```

```{r}
plotTSNE(secretory[,secretory$clincluster_final %in% c("C3","C4","C8","C10","C9")], colour_by = "Cell_subtype") +
    # theme(legend.position = "top" )  + 
    xlab("TSNE_1") + ylab("TSNE_2")
# ggsave("../../revision_analysis_20190827/revision_plots/main/secretory_TSNE_colSubtype20191007.png", width=5, height = 3.5)
```

```{r}
plotUMAP(secretory[,secretory$clincluster_final %in% c("C3","C4","C8","C10")], colour_by  = "Cell_subtype")
```


#### UMAP

Plot UMAP of all fresh cells

```{r}
# ciliated.markers <- read.csv("../tables/20180725allFT_clincluster_markers.csv", as.is = T, row.names = 1)
# ciliated.markers <- ciliated.markers$gene[ciliated.markers$logFC > 5 & ciliated.markers$cluster == 11]
fresh <- sceset[,sceset$source == "Fresh"]

# hvg only in clude markers of the four secretory cells and ciliated markers
# feature.set <- c(markers2$gene[markers2$cluster %in% c("C10","C3","C4","C8")], ciliated.markers)
# set.seed(14567)
# fresh <- runUMAP(fresh, ncomponents = 2, feature_set = rownames(fresh) %in% feature.set,
#    exprs_values = "logcounts", scale_features = TRUE)

fresh$cell_type <- NA
fresh$cell_type[colnames(fresh) %in% colnames(sceset)[sceset$final.clusters == 11]] <- "Ciliated"
fresh$cell_type[colnames(fresh) %in% colnames(secretory)] <- "Secretory"

table(fresh$cell_type)
```

```{r}
plotUMAP(fresh, colour_by = "Patient") + xlab("UMAP_1") + ylab("UMAP_2")
# p2 <- plotUMAP(fresh, 
#          colour_by = "Patient2") + xlab("UMAP_1") + ylab("UMAP_2")
# cowplot::plot_grid(p1,p2)
# ggsave("exprloratory_plots/20190508UMAP_fresh_only.png", height = 3, width = 6)
```


```{r}
plotUMAP(fresh, colour_by = "CAPS") + xlab("UMAP_1") + ylab("UMAP_2")
```


```{r}
plotUMAP(fresh[, fresh$Patient2 == 11545 & fresh$cell_type %in% c("Differentiated (C3)","KRT17 cluster (C4)","Cell cycle (C10)","EMT (C8)","Ciliated")], 
         colour_by = "cell_type") + xlab("UMAP_1") + ylab("UMAP_2")
```


```{r}
plotUMAP(fresh[, fresh$cell_type %in% c("C3","C4","C8","C10","Ciliated")], colour_by = "JUN") + xlab("UMAP_1") + ylab("UMAP_2")
```


#### Heatmap




```{r, fig.height=5}
# secretory$clincluster_final <- paste("C",secretory$clincluster_final, sep = "")
    
top10 <- markers2 %>% group_by(cluster) %>% top_n(15, logFC)
top10 <- top10[!top10$cluster %in% c("C1","C2","C5") ,]
top10 <- top10[top10$logFC > 1.8, ]

top10 <- top10[order(top10$cluster, decreasing = F),]

plot.data <- logcounts(secretory)[c(top10$gene), order(secretory$clincluster_final, decreasing = F)]

colanno <- data.frame (colData(secretory)[,c("clincluster_final","Patient2")])
colnames(colanno)[1] <- "clusters"
colanno$clusters <- factor(colanno$clusters)

rownames(colanno) <- colnames(secretory)
colanno <- colanno[order(colanno$clusters, decreasing = F),]
colanno$clusters <- factor(colanno$clusters, levels = unique(colanno$clusters))
plot.data <- plot.data[,match(rownames(colanno), colnames(plot.data))] 

plot.data <- t(scale(t(plot.data), center = T, scale = T))
# range(plot.data )
plot.data <- Seurat::MinMax(plot.data, min = -2.5, max = 2.5)

plot.data<- as.data.frame(x = t(x = plot.data))
plot.data$cell <- rownames(x = plot.data)

cells.ident <- secretory$clincluster_final
names(x = cells.ident) <- secretory$Sample

colnames(x = plot.data) <- make.unique(names = colnames(x = plot.data))
plot.data %>% melt(id.vars = "cell") -> plot.data
names(x = plot.data)[names(x = plot.data) == 'variable'] <- 'gene'
names(x = plot.data)[names(x = plot.data) == 'value'] <- 'expression'
plot.data$ident <- cells.ident[plot.data$cell]

plot.data$gene <- with(
    data = plot.data,
    expr = factor(x = gene, levels = rev(x = unique(x = plot.data$gene)))
    )
plot.data$cell <- with(
    data = plot.data,
    expr = factor(x = cell, levels = unique(x = colnames(secretory)))
    )

plot.data$ident <- factor(plot.data$ident, levels = c("C1","C2","C10","C3","C4","C5","C6","C8","C9"))
# http://colorbrewer2.org/#type=sequential&scheme=Blues&n=6

heatmap <- ggplot( data = plot.data, mapping = aes(x = cell, y = gene, fill = expression)) + 
    geom_tile() +
    scale_fill_gradient2( # colour
        low = muted("steelblue4"), mid = "white",
        high = muted("firebrick2")
        # low = "purple", mid = "black",
        # high = "yellow"
        ) +
    scale_y_discrete(position = "right", labels = rev(top10$gene)) +
    theme(
      axis.line = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank(),
      strip.text.x = element_text(size = 12),
      axis.text.y = element_text(size=6, face="italic"),
      axis.text.x = element_blank(),
      axis.title.x = element_blank(),
      strip.background = element_blank()
    )

heatmap <- heatmap +
      facet_grid(
        facets = ~ident,
        drop = TRUE,
        space = "free",
        scales = "free",
        switch = 'x'
      ) +
      scale_x_discrete(expand = c(0, 0), drop = TRUE)  
      
panel.spacing <- unit(x = 0.15, units = 'lines')
heatmap <- heatmap +
      theme(panel.spacing = panel.spacing)

heatmap
```

"The other protein of great interest, MCM2, is one of six proteins (MCM 2â€“7) that comprise the MCM helicase complex involved in the regulation of the S phase of the cell cycle in eukaryotes via the initiation of DNA replication (Kang et al, 2014). High MCM2 expression is reportedly a poor prognostic factor in ovarian cancer (Gakiopoulou et al, 2007). Furthermore, another member of the MCM2 family, MCM7, is reportedly a particularly poor prognostic factor in ovarian HGSC (Ota et al, 2011). In the present study, MCM2 was found to be differentially expressed in tubal and peritoneal HGSCs in addition to ovarian and endometrial HGSCs. "  (https://www.nature.com/articles/bjc201627)



```{r}
## Add colour bar
# plot.data$patient <- secretory$Patient2[match(plot.data$cell, colnames(secretory))]
# ggplot(plot.data, aes(x = cell, y = 1, fill = ident)) + geom_tile() + theme_void() +
#       facet_grid(
#         facets = ~ident,
#         drop = TRUE,
#         space = "free",
#         scales = "free",
#         switch = 'x'
#       ) +
#       theme(panel.spacing = panel.spacing)

```

```{r}
ggplot(plot.data, aes(x = cell, y = 1, fill = patient)) + geom_tile() + theme_void() + #theme_classic() +
      facet_grid(
        facets = ~ident,
        drop = TRUE,
        space = "free",
        scales = "free",
        switch = 'x'
      ) +
      theme(panel.spacing = panel.spacing#, 
            # axis.text.x = element_text(size = 3, angle = 90)
            ) 
# ggsave("../../revision_analysis_20190827/revision_plots/main/Fig2B_bar.pdf", width = 12, height = 3)
```


```{r}
# ggsave(filename = "../manuscript_plots/secretory subtypes/Figure_secretory_heatmap_9clusters_yellowPurple.pdf", height = 18, width = 24, units = "cm" )
# ggsave(filename = "../manuscript_plots/secretory subtypes/Figure_secretory_heatmap_9clusters.png",dpi = 300, height = 18, width = 24, units = "cm" )
```

```{r}
table(secretory$clincluster_final, secretory$Patient2 )
```

#### Violin plots of markers for cycling cluster

```{r}
secretory$clincluster_final  <- as.factor(secretory$clincluster_final)

share.7a3.1 <- c( "MKI67","PCNA","CDK4","CDC45",
             "FANCD2","FANCI","MSH2","MSH6",
             "RUVBL2","RUVBL1","HMGB2","SMC1A")

plotExpression(secretory[,secretory$clincluster_final %in% c("C10","C3","C4","C6","C8")], 
               share.7a3.1,
               x = "clincluster_final", exprs_values = "logcounts", ncol = 4, xlab = "") + 
    scale_x_discrete(labels = list("C10"="CC","C3"="C3","C4"="KRT17","C8"="ECM")) + 
    theme(strip.text = element_text(face = "italic", size = 9))
          # axis.text.x = element_text(angle = 25, vjust = 0.5, hjust = 0)  )

# ggsave("../manuscript_plots/secretory subtypes/marker_cell_cycle_violin.pdf", height = 8, width = 12, units = "cm", dpi = 300)

# ggsave("../manuscript_plots/secretory subtypes/marker_cell_cycle_violin.png", height = 8, width = 12, units = "cm", dpi = 300)
```

#### Violin plots

```{r}
share.7a3.1 <- c( "MKI67","PCNA","CDK4","CDC45",
             "FANCD2","FANCI","MSH2","MSH6",
             "RUVBL2","RUVBL1","HMGB2","SMC1A")


plotExpression(secretory[,secretory$clincluster_final %in% c("C10","C3","C4","C6","C8")], 
              share.7a3.1,
                x = "clincluster_final", exprs_values = "normcounts",
                ncol = 4, xlab = "") + 
    theme(strip.text = element_text(size = 10, face = "italic"))

# ggsave("../manuscript_plots/secretory subtypes/marker_cell_cycle_violin.png", height = 8, width = 12, units = "cm", dpi = 300)
```


Cluster 4 is characterized by upregulation of major histocompatibility complex (MHC) Class II genes (e.g. HLA-DQA1, HLA-DPA1 and HLA-DPB1), cytokeratins (KRT17, KRT5 and KRT23) and aldehyde dehydrogenases (e.g. ALDH1A1 and ALDH3B2) 

```{r}
secretory <- normalize(secretory)
assay(secretory, "normcounts") <- removeBatchEffect(assay(secretory, "normcounts"), covariates = secretory$total_features)
assay(secretory, "normcounts")[assay(secretory, "normcounts") <0] <- 0

# sseC4_marker <- c("HLA-DQA1", "HLA-DPA1", "HLA-DPB1",
#                "KRT17", "KRT5", "KRT23",
#                "ALDH1A1", "ALDH3B2","CDKN1A")
sseC4_marker <- c("HLA-DQA1", "HLA-DPA1",# "HLA-DPB1",
               "KRT17",  "KRT23",
               "ALDH1A1", "ALDH3B2")
plotExpression(secretory[,secretory$clincluster_final %in% c("C10","C3","C4","C8")], 
               sseC4_marker,
                x = "clincluster_final", exprs_values = "logcounts",
                ncol = 2, xlab = "") + ylab("Scaled expression") + 
    theme(strip.text = element_text(size = 10, face = "italic"))

# ggsave("plots/SuppFig3_marker_C4_20190508.png", width = 4, height = 4, dpi = 300)
```


```{r}
plotExpression(secretory[,secretory$clincluster_final %in% c("C10","C3","C4","C6","C8")], 
               sseC4_marker,
                x = "clincluster_final", exprs_values = "logcounts",
                ncol = 3, xlab = "") + 
    theme(strip.text = element_text(size = 10, face = "italic"))

```


```{r}
secretory$type <- NA
secretory$type[secretory$clincluster_final == "C10"] <- "Cell cycle"
secretory$type[secretory$clincluster_final == "C3"] <- "Differentiated"
secretory$type[secretory$clincluster_final == "C4"] <- "KRT17"
secretory$type[secretory$clincluster_final == "C8"] <- "EMT"
secretory$type <- factor(secretory$type, levels = c("Cell cycle", "Differentiated", "KRT17","EMT"))
plotExpression(secretory[,secretory$clincluster_final %in% c("C10","C3","C4","C8")], 
               c("EPCAM","KRT7","SPARC","RGS16"),
                x = "type", exprs_values = "logcounts",
                ncol = 2, xlab = "") + 
    theme(strip.text = element_text(size = 10, face = "italic"),
          axis.text.x = element_text(angle = 30, vjust = 0.7))
# ggsave("../../revision_analysis_20190827/revision_plots/main/Fig3a_marker_EMT_Epcam_KRT7.pdf", width = 6, height = 4, dpi = 300)
# ggsave("plots/SuppFig3_marker_EMT_Epcam_KRT7.png", width = 6, height = 4, dpi = 300)
```


```{r}
plotPhenoData(secretory, x = "clincluster_final", y = "total_features" )  + geom_hline(yintercept = 7500, col  = "grey50") + geom_hline(yintercept = 1200, col = "grey50") + xlab("Clusters")
# ggsave("plots/SuppFig3_marker_EMT_Epcam_KRT7.png", width = 6, height = 4, dpi = 300)
# ggsave("plots/SuppFig3_total_feautres.png", width = 5, height = 4)
```


```{r}
set.seed(1234)
secretory <- runUMAP(secretory, ncomponents = 2, feature_set = markers2$gene[markers2$cluster %in% c(3,4,6,7)],
  exprs_values = "logcounts", scale_features = TRUE)

plotUMAP(secretory[,secretory$clincluster.7clusters%in%c(3,4,6,7)& secretory$Patient2 == 11553], colour_by = "clincluster.7clusters") + xlab("UMAP_1") + ylab("UMAP_2")
```


## Compare to sc16

```{r}
sc16 <- readRDS("../../20190216scRNA-seq16/clean_data/sceset_sc16.rds")
stroma <- sc16[,logcounts(sc16)["EPCAM",] == 0 & logcounts(sc16)["PTPRC",] == 0]
# write.csv(as.data.frame(stroma@colData), "../GEO/sc16_colData.csv")
# write.table(counts(stroma), "../GEO/sc16_counts.txt", quote = F, sep = "\t")
dim(stroma)
```

```{r}
matrix <- counts(secretory)[,secretory$clincluster_final == "C8"]
matrix <- cbind(matrix,counts(stroma)[match(rownames(matrix), rownames(stroma)),] )
matrix <- cpm(matrix)
dge <- edgeR::DGEList(counts = matrix[rowSums(matrix) > 5, ])
info <- c(rep("C8", 40), rep("str",91))
design <- model.matrix(~ 0 + info)
v <- voom(dge, design, plot = F)
fit <- lmFit(v, design) # Linear Model for Series of Arrays
cont.matrix <- makeContrasts(contrasts = "infoC8-infostr",levels=design)
fit <- contrasts.fit(fit, cont.matrix ) # Linear Model for Series of Arrays
fit <- eBayes(fit)

C8.m <- topTable(fit, p.value = 0.05, number = Inf, coef = 1, lfc = 0.6, sort.by = "logFC")
C8.m$gene <- rownames(C8.m)
which(markers2$gene[markers2$cluster == "C8"] %in% C8.m$gene[C8.m$logFC > 0])
```

```{r}
matrix <- counts(secretory)[,secretory$clincluster_final == "C8"]
matrix <- cbind(matrix, counts(secretory)[,secretory$clincluster_final %in% c("C3","C4","C10")])
matrix <- cbind(matrix,counts(stroma)[match(rownames(matrix), rownames(stroma)),] )
info <- c(rep("C8",sum(secretory$clincluster_final == "C8")),
           rep("secretory", sum(secretory$clincluster_final %in% c("C3","C4","C10"))),
          rep("str",91))
matrix <- cpm(matrix)
dge <- edgeR::DGEList(counts = matrix[rowSums(matrix) > 5, ])

design <- model.matrix(~ 0 + info)
v <- voom(dge, design, plot = F)
fit <- lmFit(v, design) # Linear Model for Series of Arrays
cont.matrix <- makeContrasts(contrasts = c("infoC8-infosecretory","infostr-infosecretory","infostr-infoC8"),levels=design)
fit <- contrasts.fit(fit, cont.matrix) # Linear Model for Series of Arrays
fit <- eBayes(fit)

StrvsSec.m <- topTable(fit, p.value = 0.05, number = Inf, coef = 2, lfc = 0.6, sort.by = "logFC")
StrvsSec.m$gene <- rownames(StrvsSec.m)

C8vsSec.m <- topTable(fit, p.value = 0.05, number = Inf, coef = 1, lfc = 0.6, sort.by = "logFC")
C8vsSec.m$gene <- rownames(C8vsSec.m)

StrvsC8.m <- topTable(fit, p.value = 0.05, number = Inf, coef = 1, lfc = 0.6, sort.by = "logFC")
StrvsC8.m$gene <- rownames(StrvsC8.m)
```

```{r}
stroma_control <- SingleCellExperiment(assays = list(counts = matrix ), colData = data.frame(Type = info))
logcounts(stroma_control) <- log1p(calculateCPM(stroma_control))

plotExpression(stroma_control, features = c(markers2$gene[markers2$cluster == "C8"][which(markers2$gene[markers2$cluster == "C8"] %in% C8.m$gene[C8.m$logFC > 0])]), x = "Type", ncol = 4, scales = "free")
```

```{r}
plotExpression(stroma_control, features = c("COL1A2","COL3A1"), x = "Type", ncol = 2, scales = "free")  + 
    scale_x_discrete(labels = c("EMT","Other FTESCs","Stroma"),breaks = c("C8","secretory","str")) + 
    theme(strip.background = element_rect(fill = "white"),strip.text.x = element_text(face = "italic", size = 12))
# ggsave("plots/SuppFig3_EMT_stroma_col1a.png", height = 3, width = 5)
```




```{r}
plotExpression(stroma_control, features = c("SPARC","RGS16","COL1A2","COL3A1","EPCAM","KRT7"), 
               x = "Type", ncol = 2, scales = "free") + 
    scale_x_discrete(labels = c("EMT","Other FTESCs","Stroma"),breaks = c("C8","secretory","str")) + 
    theme(strip.background = element_rect(fill = "white"),strip.text.x = element_text(face = "italic", size = 12))

# ggsave("plots/SuppFig3_EMT_markers_with_stroma_control.png", width = 6, height = 8)
```


MIR17HG: This gene is the host gene for the MIR17-92 cluster, a group of at least six microRNAs (miRNAs) that may be involved in cell survival, proliferation, differentiation, and angiogenesis. Amplification of this gene has been found in several lymphomas and solid tumors. Two non-protein coding transcript variants have been found for this host gene, but only the longest is a polycistronic transcript containing the MIR17-92 cluster. [provided by RefSeq, May 2012]

RHOB: Also required for stability and nuclear trafficking of AKT1/AKT which promotes endothelial cell survival during vascular development. Serves as a microtubule-dependent signal that is required for the myosin contractile ring formation during cell cycle cytokinesis. Required for genotoxic stress-induced cell death in breast cancer cells.

#### DoubleFinder

```{r}
fresh_seu <- Seurat::as.seurat(fresh)
sc16_seu <- Seurat::as.seurat(sc16)
sc16_seu@meta.data$Source <- "Fresh"
fresh_seu@meta.data <- fresh_seu@meta.data[,colnames(fresh_seu@meta.data) %in% colnames(sc16_seu@meta.data)]
sc16_seu@meta.data <- sc16_seu@meta.data[,match(colnames(fresh_seu@meta.data), colnames(sc16_seu@meta.data))]

fresh_seu@meta.data <- fresh_seu@meta.data [,c(1,2,3,4,5,21)]
sc16_seu@meta.data <- sc16_seu@meta.data [,c(1,2,3,4,5,21)]

seurat <- Seurat::MergeSeurat(object1 = fresh_seu, object2 =  sc16_seu)
```

```{r}
seurat <- FindVariableGenes(seurat)
seurat <- ScaleData(seurat)
seurat <- RunPCA(seurat, pc.genes =  seurat@var.genes)

fresh2 <- as.SingleCellExperiment(seurat)
set.seed(12334);fresh2 <- runTSNE(fresh2)
plotTSNE(fresh2, colour_by  = "PTPRC")
```


```{r}
library(DoubletFinder)
```


```{r}
seurat <- NormalizeData(seurat)
seurat <- FindVariableGenes(seurat, x.low.cutoff = 0.0125, y.cutoff = 0.25, do.plot=FALSE)
seurat <- ScaleData(object = seurat, genes.use = seurat@var.genes)
seurat <- RunPCA(seurat, pc.genes = seurat@var.genes, pcs.print = 0)
seurat <- RunTSNE(seurat, dims.use = 1:10, verbose=TRUE)
DimElbowPlot(seurat)
```

```{r}
seurat <- FindClusters(object = seurat, reduction.type = "pca", dims.use = 1:10, 
    resolution = 0.3, print.output = 0, save.SNN = TRUE, force.recalc = T)
TSNEPlot(object = seurat)
```


```{r}
VlnPlot(object = seurat, features.plot = c("CAPS"))
```


```{r}
## pK Identification -
sweep.res.list <- paramSweep(seurat, PCs = 1:10)
sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)
bcmvn <- find.pK(sweep.stats) #0.07
```



```{r}
## Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------
annotations <- seurat@ident
homotypic.prop <- modelHomotypic(annotations)           ## ex: annotations <- seu_kidney@meta.data$ClusteringResults
nExp_poi <- round(0.01*length(seurat@cell.names))  ## 1% based on 2000 cells 
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
```


```{r}
## Run DoubletFinder with varying classification stringencies ----------------------------------------------------------------
seurat <- doubletFinder(seurat, PCs = 1:10, pN = 0.25, pK = 0.07, nExp = nExp_poi, reuse.pANN = FALSE)
seurat <- doubletFinder(seurat, PCs = 1:10, pN = 0.25, pK = 0.07, nExp = nExp_poi.adj, reuse.pANN = "pANN_0.25_0.07_25")
```


```{r}
## Plot results --------------------------------------------------------------------------------------------------------------
seurat@meta.data$DF_hi.lo <- seurat@meta.data$DF.classifications_0.25_0.07_25
seurat@meta.data$DF_hi.lo[which(seurat@meta.data$DF_hi.lo == "Doublet" & seurat@meta.data$DF.classifications_0.25_0.07_21 == "Singlet")] <- "Doublet_lo"
seurat@meta.data$DF_hi.lo[which(seurat@meta.data$DF_hi.lo == "Doublet")] <- "Doublet_hi"

TSNEPlot(seurat, group.by="DF_hi.lo", plot.order=c("Doublet_hi","Doublet_lo","Singlet"), colors.use=c("black","gold","red"))
```

```{r}
seurat@meta.data$cell_type <- NA
seurat@meta.data$cell_type <- fresh$cell_type[match(rownames(seurat@meta.data), colnames(fresh))]
seurat@meta.data$cell_subtype <- secretory$cell_type[match(rownames(seurat@meta.data), colnames(secretory))]

table(seurat@meta.data$cell_subtype, seurat@meta.data$DF_hi.lo)
```

```{r}
# saveRDS(seurat,"rds/20190508Seurat_all_fresh_doublefinder.rds", compress = T)
```


## Technical

```{r}
sessionInfo()
```